File: ToDoModel/TreeNodeOfT.vb
Public Class TreeNode(Of T)
    'Public ID As String

    Public Sub New(ByVal value As T)
        Me.Value = value
    End Sub

    Default Public ReadOnly Property Item(ByVal i As Integer) As TreeNode(Of T)
        Get
            Return Children(i)
        End Get
    End Property


    Public Property Parent As TreeNode(Of T)

    Public ReadOnly Property Value As T

    Public Function IsAncestor(ByVal model As TreeNode(Of T)) As Boolean
        If Me Is model Then Return True
        If Parent Is Nothing Then Return False
        Return Parent.IsAncestor(model)
    End Function

    Public ReadOnly Property ChildCount As Integer
        Get
            Return Children.Count
        End Get
    End Property


    Public Property Children As New List(Of TreeNode(Of T))()


    Public Function AddChild(ByVal value As T) As TreeNode(Of T)
        Dim node = New TreeNode(Of T)(value) With {
            .Parent = Me
            }
        'node.ID = NextChildID()
        Children.Add(node)
        Return node
    End Function
    Public Function AddChild(ByVal node As TreeNode(Of T)) As TreeNode(Of T)
        'node.Parent = Me
        'node.ID = NextChildID()
        Children.Add(node)
        Return node
    End Function

    Public Function InsertChild(ByVal node As TreeNode(Of T)) As TreeNode(Of T)
        node.Parent = Me
        'node.ID = strID
        Children.Insert(0, node)
        Return node
    End Function
    Public Function AddChild(ByVal value As T, ByVal strID As String) As TreeNode(Of T)
        Dim node = New TreeNode(Of T)(value) With {
                .Parent = Me
            }
        'node.ID = strID
        Children.Add(node)
        Return node
    End Function
    Public Function AddChildren(ParamArray values As T()) As TreeNode(Of T)()
        Return values.[Select](New Func(Of T, TreeNode(Of T))(AddressOf AddChild)).ToArray()
    End Function

    Public Function RemoveChild(ByVal node As TreeNode(Of T)) As Boolean
        Return Children.Remove(node)
    End Function

    Public Sub Traverse(ByVal action As Action(Of T))
        action(Value)

        For Each child In Children
            child.Traverse(action)
        Next
    End Sub

    Public Sub Traverse(ByVal action As Action(Of TreeNode(Of T)))
        action(Me)

        For Each child In Children
            child.Traverse(action)
        Next
    End Sub

    Public Function FindByDelegate(comparator As Func(Of T, String, Boolean), StringToCompare As String)
        Dim node As TreeNode(Of T)

        For Each node In Children
            If comparator(Value, StringToCompare) Then
                Return node
            End If
        Next
        Return Nothing
    End Function

    Public Function FindByDelegate(comparator As Func(Of T, T, Boolean), T2 As T)
        Dim node As TreeNode(Of T)

        For Each node In Children
            If comparator(Value, T2) Then
                Return node
            End If
        Next
        Return Nothing
    End Function

    Public Function Flatten() As IEnumerable(Of T)
        Return {Value}.Concat(Children.SelectMany(Function(x) x.Flatten()))
    End Function
End Class

File: ToDoModel/TreeNodeOfT.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace ToDoModel
{
    public class TreeNode<T>
    {
        // Public ID As String

        public TreeNode(T value)
        {
            Value = value;
        }

        public TreeNode<T> this[int i]
        {
            get
            {
                return Children[i];
            }
        }


        public TreeNode<T> Parent { get; set; }

        public T Value { get; private set; }

        public bool IsAncestor(TreeNode<T> model)
        {
            if (ReferenceEquals(this, model))
                return true;
            if (Parent is null)
                return false;
            return Parent.IsAncestor(model);
        }

        public int ChildCount
        {
            get
            {
                return Children.Count;
            }
        }


        public List<TreeNode<T>> Children { get; set; } = new List<TreeNode<T>>();


        public TreeNode<T> AddChild(T value)
        {
            var node = new TreeNode<T>(value) { Parent = this };
            // node.ID = NextChildID()
            Children.Add(node);
            return node;
        }
        public TreeNode<T> AddChild(TreeNode<T> node)
        {
            // node.Parent = Me
            // node.ID = NextChildID()
            Children.Add(node);
            return node;
        }

        public TreeNode<T> InsertChild(TreeNode<T> node)
        {
            node.Parent = this;
            // node.ID = strID
            Children.Insert(0, node);
            return node;
        }
        public TreeNode<T> AddChild(T value, string strID)
        {
            var node = new TreeNode<T>(value) { Parent = this };
            // node.ID = strID
            Children.Add(node);
            return node;
        }
        public TreeNode<T>[] AddChildren(params T[] values)
        {
            return values.Select(new Func<T, TreeNode<T>>(AddChild)).ToArray();
        }

        public bool RemoveChild(TreeNode<T> node)
        {
            return Children.Remove(node);
        }

        public void Traverse(Action<T> action)
        {
            action(Value);

            foreach (var child in Children)
                child.Traverse(action);
        }

        public void Traverse(Action<TreeNode<T>> action)
        {
            action(this);

            foreach (var child in Children)
                child.Traverse(action);
        }

        public object FindByDelegate(Func<T, string, bool> comparator, string StringToCompare)
        {

            foreach (var node in Children)
            {
                if (comparator(Value, StringToCompare))
                {
                    return node;
                }
            }
            return null;
        }

        public object FindByDelegate(Func<T, T, bool> comparator, T T2)
        {

            foreach (var node in Children)
            {
                if (comparator(Value, T2))
                {
                    return node;
                }
            }
            return null;
        }

        public IEnumerable<T> Flatten()
        {
            return new[] { Value }.Concat(Children.SelectMany(x => x.Flatten()));
        }
    }
}